tests/unit/s2n_self_talk_npn_test.c

```diff
@@ -0,0 +1,176 @@
+ * hello and forces the server to negotiate a protocol using the NPN extension 
+ * instead. */
+static int s2n_wipe_alpn_ext(struct s2n_connection *conn, void *ctx)
+{
+    struct s2n_client_hello *client_hello = s2n_connection_get_client_hello(conn);
+    POSIX_ENSURE_REF(client_hello);
+    s2n_parsed_extension *parsed_extension = NULL;
+    POSIX_GUARD(s2n_client_hello_get_parsed_extension(S2N_EXTENSION_ALPN, &client_hello->extensions, &parsed_extension));
+    POSIX_ENSURE_REF(parsed_extension);
+    POSIX_GUARD(s2n_blob_zero(&parsed_extension->extension));
+
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    const char *protocols[] = { "http/1.1", "spdy/1", "spdy/2" };
+    const uint8_t protocols_count = s2n_array_len(protocols);
+
+    /* Set up connections */
+    DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+    EXPECT_NOT_NULL(server_conn);
+    DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+    EXPECT_NOT_NULL(client_conn);
+
+    /* Set up config */
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+    EXPECT_NOT_NULL(config);
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+    struct s2n_cert_chain_and_key *chain_and_key = NULL;
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key, S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+    EXPECT_SUCCESS(s2n_config_set_protocol_preferences(config, protocols, protocols_count));
+
+    /* Set up config that wipes ALPN extension */
+    DEFER_CLEANUP(struct s2n_config *npn_config = s2n_config_new(), s2n_config_ptr_free);
+    EXPECT_NOT_NULL(npn_config);
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(npn_config));
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(npn_config, "default"));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(npn_config, chain_and_key));
+    EXPECT_SUCCESS(s2n_config_set_protocol_preferences(npn_config, protocols, protocols_count));
+    EXPECT_SUCCESS(s2n_config_set_client_hello_cb(npn_config, s2n_wipe_alpn_ext, NULL));
+
+    EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+    EXPECT_SUCCESS(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+
+    /* Client and server both support NPN. ALPN is negotiated since it is also
+     * supported and the server prefers ALPN. */
+    {
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        server_conn->config->npn_supported = true;
+        client_conn->config->npn_supported = true;
```

Comment:
Nit: this is true for all your tests, and involves modifying the config. You should probably just do it once when the config is created.